pkg/apis/storage/types.go

```diff
@@ -337,6 +337,47 @@ type CSIDriverSpec struct {
 	//
 	// +optional
 	StorageCapacity *bool
+
+	// ServiceAccountTokens indicates the csi driver needs pods' service account
+	// tokens it is mounting volume for to do necessary authentication. Kubelet
+	// will pass the tokens in VolumeContext in the CSI NodePublishVolume calls.
+	// The CSI driver should parse and validate the following VolumeConext:
+	// "csi.storage.k8s.io/serviceAccount.tokens": {
+	//   '<audience>': {
+	//     'token': <token>,
+	//     'expiry':<token>,
+	//   },
+	//   ...
+	//   // If audience is not specified, the output key will be ''
+	//   '': {
+	//     'token': <token>,
+	//     'expiry': <expiry>,
+	//   },
+	// }
+	//
+	// Note: Audience in each ServiceAccountToken should be different and at
+	// most one token can ignore Audience.
+	//
+	// This is an alpha feature and only available when the
+	// CSIServiceAccountToken feature is enabled.
+	//
+	// +optional
+	// +listType=set
```

Comment:
Yeah, that's questionnable. I believe this should be mechanically forbidden.
We don't support them yet in SSA, I don't think we've seen any case where this was needed.
So yes, the question is:
> what is listType=set accomplishing here?

Reply:
oh, it is not supported in SSA? reading from https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/0006-apply.md#api-topology that's the way to specify object topology.

Reply:
non-primitive types aren't supported yet. We could probably make this work if needed but I'm curious if this is actually doing what you want to do. What are you trying to achieve?

Reply:
Actually, doc is correct, it reads:
> Sets: the list is a set (it has to be of a scalar type). [snipped...]

Reply:
this list of ServiceAccountToken should be unique. not sure whether we can validate from API side or this marker is just informational.

Reply:
If I read properly, `ServiceAccountToken` is made of an optional `Audience` and 
 another optional `ExpirationSeconds`. What is supposed to be unique? Can there be two entries with the same Audience? Or are you trying to have unique Audience/ExpirationSeconds combination?

I still find it hard to reason about this because i find the name `ServiceAccountToken` to be misleading. It's very likely that I'm just misunderstanding, but the naming doesn't help me here :-)

>	// Note: Audience in each ServiceAccountToken should be different and at
>	// most one token can ignore Audience.

Based on that, I understand that you want `ServiceAccountTokens` to be an associative-list (a map) and the key should be `Audience`. We don't allow optional non-defaulted fields to be keys, but you could make audience required and use empty-string for "ignore", would that work?

Reply:
Also note that while this is not strictly informational (it's used for merging on server-side apply), at this point you still have to write manual validation for that unfortunately.

Reply:
so if using `+patchMergeKey=name`, the audience has to be a required `string` instead of an optional `*string`? wrt naming, probably `TokenRequest` more accurate.

Reply:
Slightly unclear, you don't have a `name` field here, so I assume you mean `+patchMergeKey=audience`. In that case, yes audience is required. I'm working on a change that would allow you to use a defaulted empty string for audience while still work as the patchMergeKey.

Reply:
yea, i meant `audience` key. the change would be needed by this PR. when do you the change would be ready?

Reply:
Sorry I missed that comment, as answered privately, we're planning on having that in 1.20.

Reply:
thank you. for now i'll use `listType=atomic` until the changes land in 1.20. PTAL.